File: packages/desktop/Sources/Services/APIClient.swift
import Foundation

// MARK: - API Error

enum CortexAPIError: LocalizedError {
    case invalidURL
    case networkError(Error)
    case httpError(statusCode: Int, message: String?)
    case decodingError(Error)
    case daemonOffline
    case unknown(String)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .httpError(let code, let message):
            return "HTTP \(code): \(message ?? "Unknown error")"
        case .decodingError(let error):
            return "Failed to decode response: \(error.localizedDescription)"
        case .daemonOffline:
            return "Cortex daemon is not running"
        case .unknown(let message):
            return message
        }
    }
}

// MARK: - API Client

actor APIClient {
    static let shared = APIClient()

    private let baseURL = "http://127.0.0.1:7434"
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        config.timeoutIntervalForResource = 30
        config.waitsForConnectivity = false
        self.session = URLSession(configuration: config)

        self.decoder = JSONDecoder()
        self.decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let dateString = try container.decode(String.self)

            // Try ISO8601 with fractional seconds
            let iso8601Formatter = ISO8601DateFormatter()
            iso8601Formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let date = iso8601Formatter.date(from: dateString) {
                return date
            }

            // Try ISO8601 without fractional seconds
            iso8601Formatter.formatOptions = [.withInternetDateTime]
            if let date = iso8601Formatter.date(from: dateString) {
                return date
            }

            // Try Unix timestamp
            if let timestamp = Double(dateString) {
                return Date(timeIntervalSince1970: timestamp)
            }

            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Cannot decode date: \(dateString)"
            )
        }

        self.encoder = JSONEncoder()
        self.encoder.keyEncodingStrategy = .convertToSnakeCase
    }

    // MARK: - Health

    func health() async throws -> HealthResponse {
        return try await get("/api/health")
    }

    func isHealthy() async -> Bool {
        do {
            let response: HealthResponse = try await get("/api/health")
            return response.isHealthy
        } catch {
            return false
        }
    }

    // MARK: - Projects

    func listProjects() async throws -> ProjectListResponse {
        return try await get("/api/projects")
    }

    func getProject(id: String) async throws -> Project {
        return try await get("/api/projects/\(id)")
    }

    // MARK: - Memories

    func listMemories(
        projectId: String? = nil,
        type: MemoryType? = nil,
        limit: Int = 50,
        page: Int = 1
    ) async throws -> MemoryListResponse {
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "limit", value: String(limit)),
            URLQueryItem(name: "page", value: String(page))
        ]
        if let projectId = projectId {
            queryItems.append(URLQueryItem(name: "project_id", value: projectId))
        }
        if let type = type {
            queryItems.append(URLQueryItem(name: "type", value: type.rawValue))
        }
        return try await get("/api/memories", queryItems: queryItems)
    }

    func getMemory(id: String) async throws -> Memory {
        return try await get("/api/memories/\(id)")
    }

    func createMemory(body: MemoryCreateRequest) async throws -> Memory {
        return try await post("/api/memories", body: body)
    }

    func updateMemory(id: String, body: MemoryUpdateRequest) async throws -> Memory {
        return try await patch("/api/memories/\(id)", body: body)
    }

    func deleteMemory(id: String) async throws {
        let _: EmptyResponse = try await delete("/api/memories/\(id)")
    }

    func searchMemories(query: String, type: MemoryType? = nil) async throws -> SearchResponse {
        var queryItems = [URLQueryItem(name: "q", value: query)]
        if let type = type {
            queryItems.append(URLQueryItem(name: "type", value: type.rawValue))
        }
        return try await get("/api/memories/search", queryItems: queryItems)
    }

    // MARK: - Sync

    func syncStatus() async throws -> SyncStatusResponse {
        return try await get("/api/sync/status")
    }

    func syncNow() async throws -> SyncStatusResponse {
        return try await post("/api/sync/now", body: EmptyBody())
    }

    func syncSetup(url: String, token: String) async throws -> SyncStatusResponse {
        let request = SyncSetupRequest(url: url, token: token)
        return try await post("/api/sync/setup", body: request)
    }

    // MARK: - Analytics

    func getAnalytics() async throws -> AnalyticsResponse {
        return try await get("/api/analytics")
    }

    // MARK: - Config

    func getConfig() async throws -> CortexConfig {
        return try await get("/api/config")
    }

    func updateConfig(_ config: CortexConfig) async throws -> CortexConfig {
        return try await put("/api/config", body: config)
    }

    // MARK: - HTTP Methods

    private func get<T: Decodable>(_ path: String, queryItems: [URLQueryItem] = []) async throws -> T {
        let request = try buildRequest(path: path, method: "GET", queryItems: queryItems)
        return try await execute(request)
    }

    private func post<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        var request = try buildRequest(path: path, method: "POST")
        request.httpBody = try encoder.encode(body)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return try await execute(request)
    }

    private func put<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        var request = try buildRequest(path: path, method: "PUT")
        request.httpBody = try encoder.encode(body)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return try await execute(request)
    }

    private func patch<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        var request = try buildRequest(path: path, method: "PATCH")
        request.httpBody = try encoder.encode(body)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return try await execute(request)
    }

    private func delete<T: Decodable>(_ path: String) async throws -> T {
        let request = try buildRequest(path: path, method: "DELETE")
        return try await execute(request)
    }

    // MARK: - Request Building

    private func buildRequest(
        path: String,
        method: String,
        queryItems: [URLQueryItem] = []
    ) throws -> URLRequest {
        guard var components = URLComponents(string: baseURL + path) else {
            throw CortexAPIError.invalidURL
        }

        if !queryItems.isEmpty {
            components.queryItems = queryItems
        }

        guard let url = components.url else {
            throw CortexAPIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        return request
    }

    private func execute<T: Decodable>(_ request: URLRequest) async throws -> T {
        let data: Data
        let response: URLResponse

        do {
            (data, response) = try await session.data(for: request)
        } catch let urlError as URLError where urlError.code == .cannotConnectToHost
            || urlError.code == .networkConnectionLost
            || urlError.code == .timedOut {
            throw CortexAPIError.daemonOffline
        } catch {
            throw CortexAPIError.networkError(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw CortexAPIError.unknown("Invalid response type")
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            let message = String(data: data, encoding: .utf8)
            throw CortexAPIError.httpError(statusCode: httpResponse.statusCode, message: message)
        }

        // Handle empty responses
        if data.isEmpty, let emptyResponse = EmptyResponse() as? T {
            return emptyResponse
        }

        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw CortexAPIError.decodingError(error)
        }
    }
}

// MARK: - Helper Types

private struct EmptyBody: Encodable {}

struct EmptyResponse: Codable {
    init() {}
}

File: packages/cli/src/api-client.ts
/**
 * HTTP client for the Cortex daemon REST API at localhost:7434.
 * All CLI commands use this instead of direct SQLite access.
 */

export class APIClient {
  private baseUrl: string;

  constructor(port = 7434) {
    this.baseUrl = `http://127.0.0.1:${port}`;
  }

  /** Check if daemon is running */
  async isHealthy(): Promise<boolean> {
    try {
      const res = await fetch(`${this.baseUrl}/api/health`, { signal: AbortSignal.timeout(2000) });
      return res.ok;
    } catch {
      return false;
    }
  }

  /** GET /api/health */
  async health() {
    return this.get<{
      status: string;
      version: string;
      db_ok: boolean;
      sync_ok: boolean;
      uptime_s: number;
      memory_count: number;
      db_size_mb: number;
      schema_version: number;
    }>('/api/health');
  }

  /** GET /api/projects */
  async listProjects() {
    return this.get<{ data: any[]; total: number }>('/api/projects');
  }

  /** GET /api/projects/:id */
  async getProject(id: string) {
    return this.get<{ data: { project: any; stats: any } }>(`/api/projects/${id}`);
  }

  /** PATCH /api/projects/:id */
  async updateProject(id: string, body: Record<string, unknown>) {
    return this.patch<{ data: any }>(`/api/projects/${id}`, body);
  }

  /** DELETE /api/projects/:id */
  async deleteProject(id: string) {
    return this.del(`/api/projects/${id}`);
  }

  /** GET /api/memories?project_id=X */
  async listMemories(params: Record<string, string>) {
    const qs = new URLSearchParams(params).toString();
    return this.get<{ data: any[]; total: number }>(`/api/memories?${qs}`);
  }

  /** GET /api/memories/:id */
  async getMemory(id: string) {
    return this.get<{ data: any }>(`/api/memories/${id}`);
  }

  /** POST /api/memories */
  async createMemory(body: Record<string, unknown>) {
    return this.post<{ data: any }>('/api/memories', body);
  }

  /** PATCH /api/memories/:id */
  async updateMemory(id: string, body: Record<string, unknown>) {
    return this.patch<{ data: any }>(`/api/memories/${id}`, body);
  }

  /** DELETE /api/memories/:id */
  async deleteMemory(id: string) {
    return this.del(`/api/memories/${id}`);
  }

  /** POST /api/memories/:id/supersede */
  async supersedeMemory(id: string, body: Record<string, unknown>) {
    return this.post<{ data: any }>(`/api/memories/${id}/supersede`, body);
  }

  /** POST /api/memories/search */
  async searchMemories(body: Record<string, unknown>) {
    return this.post<{ data: any[]; total: number }>('/api/memories/search', body);
  }

  /** GET /api/memories/stale */
  async getStaleMemories(projectId?: string) {
    const qs = projectId ? `?project_id=${projectId}` : '';
    return this.get<{ data: any[]; total: number }>(`/api/memories/stale${qs}`);
  }

  /** POST /api/memories/:id/rate */
  async rateMemory(id: string, rating: number) {
    return this.post<{ data: { ok: boolean } }>(`/api/memories/${id}/rate`, { rating });
  }

  /** GET /api/sessions?project_id=X */
  async listSessions(projectId: string, limit = 20) {
    return this.get<{ data: any[] }>(`/api/sessions?project_id=${projectId}&limit=${limit}`);
  }

  /** GET /api/analytics */
  async getAnalytics() {
    return this.get<{ data: any }>('/api/analytics');
  }

  /** GET /api/config */
  async getConfig() {
    return this.get<{ data: any }>('/api/config');
  }

  /** PATCH /api/config */
  async updateConfig(body: Record<string, unknown>) {
    return this.patch<{ data: any }>('/api/config', body);
  }

  /** POST /api/sync/start */
  async startSync() {
    return this.post<{ data: any }>('/api/sync/start', {});
  }

  /** POST /api/sync/stop */
  async stopSync() {
    return this.post<{ data: any }>('/api/sync/stop', {});
  }

  /** GET /api/sync/status */
  async syncStatus() {
    return this.get<{ data: any }>('/api/sync/status');
  }

  /** POST /api/sync/now */
  async syncNow() {
    return this.post<{ data: any }>('/api/sync/now', {});
  }

  /** POST /api/sync/setup */
  async syncSetup(body: { url: string; token: string }) {
    return this.post<{ data: any }>('/api/sync/setup', body);
  }

  /** POST /api/memories/:id/pin */
  async pinMemory(id: string) {
    return this.post<{ data: any }>(`/api/memories/${id}/pin`, {});
  }

  /** POST /api/memories/:id/unpin */
  async unpinMemory(id: string) {
    return this.post<{ data: any }>(`/api/memories/${id}/unpin`, {});
  }

  /** POST /api/summarize */
  async triggerSummarize(body?: Record<string, unknown>) {
    return this.post<{ data: any }>('/api/summarize', body || {});
  }

  /** POST /api/projects/:id/archive */
  async archiveProject(id: string) {
    return this.post<{ data: any }>(`/api/projects/${id}/archive`, {});
  }

  // ── HTTP helpers ──

  private async get<T>(path: string): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, { signal: AbortSignal.timeout(5000) });
    if (!res.ok) {
      const body = await res.json().catch(() => ({}));
      throw new APIError(res.status, body?.error?.code || 'UNKNOWN', body?.error?.message || res.statusText);
    }
    return res.json() as Promise<T>;
  }

  private async post<T>(path: string, body: unknown): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(5000),
    });
    if (!res.ok) {
      const rb = await res.json().catch(() => ({}));
      throw new APIError(res.status, rb?.error?.code || 'UNKNOWN', rb?.error?.message || res.statusText);
    }
    if (res.status === 204) return {} as T;
    return res.json() as Promise<T>;
  }

  private async patch<T>(path: string, body: unknown): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(5000),
    });
    if (!res.ok) {
      const rb = await res.json().catch(() => ({}));
      throw new APIError(res.status, rb?.error?.code || 'UNKNOWN', rb?.error?.message || res.statusText);
    }
    return res.json() as Promise<T>;
  }

  private async del(path: string): Promise<void> {
    const res = await fetch(`${this.baseUrl}${path}`, { method: 'DELETE', signal: AbortSignal.timeout(5000) });
    if (!res.ok && res.status !== 204) {
      const rb = await res.json().catch(() => ({}));
      throw new APIError(res.status, rb?.error?.code || 'UNKNOWN', rb?.error?.message || res.statusText);
    }
  }
}

export class APIError extends Error {
  constructor(
    public status: number,
    public code: string,
    message: string,
  ) {
    super(message);
    this.name = 'APIError';
  }
}
